integration/test/test_frequency_balancer.py

```diff
@@ -0,0 +1,134 @@
+available, the test aims to improve both energy and performance on an imbalanced
+workload.
+
+"""
+
+import sys
+import unittest
+import os
+from pathlib import Path
+import shutil
+from experiment import machine
+from types import SimpleNamespace
+
+import geopmpy.agent
+from yaml import load
+try:
+    from yaml import CSafeLoader as SafeLoader
+except ImportError:
+    from yaml import SafeLoader
+from collections import defaultdict
+
+from integration.test import util
+from integration.test import geopm_test_launcher
+from experiment.sst_evaluation import sst_evaluation
+from apps.arithmetic_intensity import arithmetic_intensity
+
+@util.skip_unless_do_launch()
+class TestIntegration_frequency_balancer(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        try:
+            cls._do_use_sst = geopm_test_launcher.geopmread("SST::TURBOFREQ_SUPPORT:SUPPORTED board 0")
+        except:
+            cls._do_use_sst = False
+
+    def tearDown(self):
+        if sys.exc_info() != (None, None, None):
+            TestIntegration_frequency_balancer._keep_files = True
+
+    def test_cores_rebalance(self):
+        """
+        An imbalanced application gets rebalanced by the frequency_balancer agent.
+        """
+        output_dir = Path(__file__).resolve().parent.joinpath('test_cores_rebalance_output')
+        if output_dir.exists() and output_dir.is_dir():
+            shutil.rmtree(output_dir)
+
+        mach = machine.init_output_dir(output_dir)
+
+        node_count = 1
+        ranks_per_node = mach.num_core() - mach.num_package()
+        ranks_per_package = ranks_per_node // mach.num_package()
+
+        try:
+            slow_ranks_per_package = int(geopm_test_launcher.geopmread("SST::HIGHPRIORITY_NCORES:0 package 0"))
+        except Exception:
+            slow_ranks_per_package = ranks_per_package // 4
+
+        # Configure the test application
+        app_conf = arithmetic_intensity.ArithmeticIntensityAppConf(
+            ['--slowdown=2.5',
+             f'--slow-ranks-per-imbalanced-group={slow_ranks_per_package}',
+             f'--ranks-per-imbalanced-group={ranks_per_package}',
+             '--base-internal-iterations=2',
+             '--iterations=100',
+             f'--floats={1<<22}',
+             '--benchmarks=32'],
+            mach,
+            run_type='avx2',
+            ranks_per_node=ranks_per_node)
+        experiment_args = SimpleNamespace(
+            output_dir=output_dir,
+            agent_list=None,
+            node_count=node_count,
+            trial_count=1,
+            cool_off_time=10,
+            enable_traces=False,
+            enable_profile_traces=False,
+            power_cap=None,
+        )
+
+        sst_evaluation.launch(app_conf, args=experiment_args, experiment_cli_args=[])
+
+        results = defaultdict(dict)
+        for report_path in output_dir.glob('*.report'):
+            with open(report_path) as f:
+                report = load(f, Loader=SafeLoader)
```

Comment:
This looks to be the same as io.RawReport.  Why not use it?
There's also RawReportCollection for several report files.

Reply:
`RawReportCollection` doesn't fit well here since I need to use agent-specific metadata for the test (i.e., which controls were requested and which controls were actually used based on availability). I think I can use `RawReport`